src/spec.rs

```diff
@@ -1,128 +1,372 @@
         }
 
-        let mut pos = file_length - HEADER_SIZE;
-        while pos >= search_upper_bound {
-            let mut have_signature = false;
-            reader.seek(io::SeekFrom::Start(pos))?;
-            if reader.read_u32_le()? == CENTRAL_DIRECTORY_END_SIGNATURE {
-                have_signature = true;
-                reader.seek(io::SeekFrom::Current(
-                    BYTES_BETWEEN_MAGIC_AND_COMMENT_SIZE as i64,
-                ))?;
-                let cde_start_pos = reader.seek(io::SeekFrom::Start(pos))?;
-                if let Ok(end_header) = CentralDirectoryEnd::parse(reader) {
-                    return Ok((end_header, cde_start_pos));
+        let search_upper_bound = 0;
+
+        const END_WINDOW_SIZE: usize = 512;
+
+        let sig_bytes = CENTRAL_DIRECTORY_END_SIGNATURE.to_le_bytes();
+        let finder = FinderRev::new(&sig_bytes);
+
+        let mut window_start: u64 = file_length.saturating_sub(END_WINDOW_SIZE as u64);
+        let mut window = [0u8; END_WINDOW_SIZE];
+        while window_start >= search_upper_bound {
+            /* Go to the start of the window in the file. */
+            reader.seek(io::SeekFrom::Start(window_start))?;
+
+            /* Identify how many bytes to read (this may be less than the window size for files
+             * smaller than END_WINDOW_SIZE). */
+            let end = (window_start + END_WINDOW_SIZE as u64).min(file_length);
+            let cur_len = (end - window_start) as usize;
+            debug_assert!(cur_len > 0);
+            debug_assert!(cur_len <= END_WINDOW_SIZE);
+            let cur_window: &mut [u8] = &mut window[..cur_len];
+            /* Read the window into the bytes! */
+            reader.read_exact(cur_window)?;
+
+            /* Find instances of the magic signature. */
+            for offset in finder.rfind_iter(cur_window) {
+                let cde_start_pos = window_start + offset as u64;
+                reader.seek(io::SeekFrom::Start(cde_start_pos))?;
+                /* Drop any headers that don't parse. */
+                if let Ok(cde) = Self::parse(reader) {
+                    return Ok((cde, cde_start_pos));
                 }
             }
-            pos = match pos.checked_sub(if have_signature {
-                size_of_val(&CENTRAL_DIRECTORY_END_SIGNATURE) as u64
-            } else {
-                1
-            }) {
-                Some(p) => p,
-                None => break,
-            };
+
+            /* We always want to make sure we go allllll the way back to the start of the file if
+             * we can't find it elsewhere. However, our `while` condition doesn't check that. So we
+             * avoid infinite looping by checking at the end of the loop. */
+            if window_start == search_upper_bound {
+                break;
+            }
+            debug_assert!(END_WINDOW_SIZE > mem::size_of_val(&CENTRAL_DIRECTORY_END_SIGNATURE));
```

Comment:
Since we now have a separate `Magic` wrapper struct with a fixed size, we can use `size_of()` instead of `size_of_val()`, which becomes a `const` expression. I have also added a note to look at the `static_assertions` crate to do this, which is explicitly intended for security-conscious checks.